WP-CLI

crstauf opened the pull request:
Port of [crstauf/action-scheduler-cli](https://github.com/crstauf/action-scheduler-cli) into Action Scheduler.

Closes #255.

Hope I'm not disappointed.

### Changelog

> Add - A number of new WP CLI commands have been added, making it easier to manage actions in the terminal and from scripts.

crstauf committed:
add files

crstauf committed:
eliminate code smells

crstauf committed:
add logs command

crstauf committed:
Merge branch 'trunk' into wpcli

vedanshujain commented on classes/WP_CLI/Action/Create.php:
```diff
@@ -0,0 +1,126 @@
+			$callback_args = json_decode( $callback_args, true );
+		}
+
+		$function_args = array(
```
let's add support for `uniq` and `priority` param too

vedanshujain commented on classes/WP_CLI/Action/Create.php:
```diff
@@ -0,0 +1,126 @@
+		$function_args = array_values( array_filter( $function_args ) );
+
+		try {
+			$actions_added = call_user_func_array( $function, $function_args );
```
this is not the number of actions added, but instead the action ID. I suppose there will always be only one action created in this command anyway. And would be `0` if there was no action created, we should also print_error when the returning ID is 0.

vedanshujain commented on classes/WP_CLI/Action/Cancel.php:
```diff
@@ -0,0 +1,79 @@
+		try {
+			call_user_func( $function_name, $hook, $callback_args, $group );
+		} catch ( \Exception $e ) {
+			$this->print_error( $e, $multiple );
```
note that it won't throw an exception necessarily even if cancelling failed, there could be an error caught by the try_catch in `as_unschedule_action`. It would be more reliable to either 
1. check the logs for failed action, or
2. verify that the action is indeed cancelled, or
3. modify the `as_unschedule_action` to return the result value so that we can display it reliably.

crstauf replied:
Going to rework this command a bit: realized in its current state actions could not be unscheduled by group only.

vedanshujain commented on classes/WP_CLI/Action/Delete.php:
```diff
@@ -0,0 +1,90 @@
+/**
+ * WP-CLI command: action-scheduler action delete
+ */
+class ActionScheduler_WPCLI_Action_Delete_Command extends ActionScheduler_WPCLI_Command {
```
do you think if we should support a bulk delete option as well, for example, deleting by hook name? folks may want to use this command to control number of action, in which case deleting by hook would be more useful than deleting by id.

crstauf replied:
Such a task should already be doable with something like this:

```shell
wp action-scheduler action delete $( wp action-scheduler list --hook=action_hook --format=ids )
```

vedanshujain commented on classes/WP_CLI/Action/Delete.php:
```diff
@@ -0,0 +1,90 @@
+		$this->action_ids             = array_map( 'absint', $args );
+		$this->action_counts['total'] = count( $this->action_ids );
+
+		add_action( 'action_scheduler_deleted_action', array( $this, 'action__deleted' ) );
```
we should have a check on whether `WP_CLI` is set before hooking this action.

crstauf replied:
Because of the way the `action` subcommands are being loaded, they only load in a `WP_CLI` environment, but I will add a check to the abstract class for redundancy.

vedanshujain commented on classes/WP_CLI/Action/Get.php:
```diff
@@ -0,0 +1,59 @@
+
+		foreach ( $logger->get_logs( $action_id ) as $log_entry ) {
+			if ( $only_logs ) {
+				WP_CLI::line( sprintf( '%s %s', $log_entry->get_date()->format( static::DATE_FORMAT ), $log_entry->get_message() ) );
```
nice, this seems very useful!

vedanshujain commented on classes/WP_CLI/Action/List.php:
```diff
@@ -0,0 +1,107 @@
+ */
+class ActionScheduler_WPCLI_Action_List_Command extends ActionScheduler_WPCLI_Command {
+
+	const PARAMETERS = array(
```
we should mention defaults, especially for pagination params somewhere.

crstauf replied:
```
$ wp help action-scheduler action list
...
SYNOPSIS

  wp action-scheduler action list [--<field>=<value>] [--field=<field>] [--fields=<fields>] [--format=<format>]

  Display actions based on all arguments supported by
  [as_get_scheduled_actions()][1].

  ---
  [1] https://actionscheduler.org/api/#function-reference--as_get_scheduled_actions
...
```

See https://github.com/crstauf/action-scheduler/blob/76e55cfd59282ab2a6a0f2303ff4df0e4dc1469d/classes/WP_CLI/ActionScheduler_WPCLI_Action_Command.php#L209-L213.

This follows the convention set by [the `wp post list` command](https://developer.wordpress.org/cli/commands/post/list/):

```
$ wp help post list
...
SYNOPSIS

  wp post list [--<field>=<value>] [--field=<field>] [--fields=<fields>] [--format=<format>]

  Display posts based on all arguments supported by [WP_Query()][1].
  Only shows post types marked as post by default.

  ---
  [1] https://developer.wordpress.org/reference/classes/wp_query/
...
```

vedanshujain replied:
ok, makes sense. although one change that we should make is change the order to `desc` by default. otherwise I am just seeing list of old actions and it's not immediately helpful.

crstauf replied:
> although one change that we should make is change the order to desc by default.

I will not be making this change: the behavior should match what is documented, and if a user wants to retrieve newer actions, they can use the `date` and `date_compare` parameters.

vedanshujain reviewed (commented):
Great work! I posted some initial thoughts as I work my way through the PR, let me know what you think.

crstauf committed:
add unique and priority parameters

crstauf committed:
use accurate variable names and language

crstauf committed:
improve fail detection and improve unscheduling of multiple actions

crstauf committed:
redundancy to ensure in wp cli environment

vedanshujain commented:
The code looks good, going to test around these flows in general:

- [ ] Create - async/single/cron/recurring/error
- [ ] Generate - testing for bulk count
- [ ] Get - error in action.
- [ ] Run
- [ ] List - listing in bulk
- [ ] Cancel - check for error flows too
- [ ] Delete - check for deleting in bulk + error flows

vedanshujain commented on classes/WP_CLI/Action/Delete.php:
```diff
@@ -0,0 +1,90 @@
+		);
+
+		foreach ( $this->action_ids as $action_id ) {
+			$store->delete_action( $action_id );
```
Seems like we not handling failures gracefully, passing one invalid ID crashes the whole command. I think we can wrap this in a try catch, and then display the actions that we were not able to delete:

```
> wp action-scheduler action delete 1561 163333
Deleting 2 actions  50 % [========================================================>                                                        ] 0:00 / 0:00Fatal error: Uncaught InvalidArgumentException: Unidentified action 163333: we were unable to delete this action. It may may have been deleted by another process. in /usr/src/public_html/wp-content/plugins/action-scheduler/classes/data-stores/ActionScheduler_DBStore.php:746
Stack trace:
#0 /usr/src/public_html/wp-content/plugins/action-scheduler/classes/WP_CLI/Action/Delete.php(53): ActionScheduler_DBStore->delete_action(163333)
#1 /usr/src/public_html/wp-content/plugins/action-scheduler/classes/WP_CLI/ActionScheduler_WPCLI_Action_Command.php(119): ActionScheduler_WPCLI_Action_Delete_Command->execute()
#2 [internal function]: ActionScheduler_WPCLI_Action_Command->delete(Array, Array)
#3 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Dispatcher/CommandFactory.php(100): call_user_func(Array, Array, Array)
#4 [internal function]: WP_CLI\Dispatcher\CommandFactory::WP_CLI\Dispatcher\{closure}(Array, Array)
#5 phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Dispatcher/Subcommand.php(491): call_user_func(Objec in /usr/src/public_html/wp-content/plugins/action-scheduler/classes/data-stores/ActionScheduler_DBStore.php on line 746
Error: There has been a critical error on this website.Learn more about troubleshooting WordPress. There has been a critical error on this website.
```

vedanshujain commented on classes/WP_CLI/Action/Get.php:
```diff
@@ -0,0 +1,59 @@
+		$action_id = $this->args[0];
+		$store     = \ActionScheduler::store();
+		$logger    = \ActionScheduler::logger();
+		$action    = $store->fetch_action( $action_id );
```
Same as delete, this throws a fatal for an action that does not exist. Let's handle that gracefully like `wp post` does.

vedanshujain commented on classes/WP_CLI/Action/Create.php:
```diff
@@ -0,0 +1,135 @@
+
+		// Generate schedule start if appropriate.
+		if ( ! in_array( $schedule_start, static::ASYNC_OPTS, true ) ) {
+			$schedule_start         = as_get_datetime_object( $schedule_start );
```
`as_get_datetime_object` should be wrapped in a try--catch, mentioned that we were not able to parse the date (and possibly displaying an example or two)

vedanshujain commented on classes/WP_CLI/Action/Create.php:
```diff
@@ -0,0 +1,135 @@
+/**
+ * WP-CLI command: action-scheduler action create
+ */
+class ActionScheduler_WPCLI_Action_Create_Command extends ActionScheduler_WPCLI_Command {
```
I am getting some issues using this command, seems like we are somehow setting the `function_args` to an int

```
> wp action-scheduler action create woocommerce_cleanup_draft_orders now

Fatal error: Uncaught TypeError: Argument 2 passed to ActionScheduler_Action::__construct() must be of the type array, int given, called in /usr/src/public_html/wp-content/plugins/action-scheduler/classes/ActionScheduler_ActionFactory.php on line 314 and defined in /usr/src/public_html/wp-content/plugins/action-scheduler/classes/actions/ActionScheduler_Action.php:26
Stack trace:
#0 /usr/src/public_html/wp-content/plugins/action-scheduler/classes/ActionScheduler_ActionFactory.php(314): ActionScheduler_Action->__construct('woocommerce_cle...', 10, Object(ActionScheduler_NullSchedule), '')
#1 /usr/src/public_html/wp-content/plugins/action-scheduler/functions.php(51): ActionScheduler_ActionFactory->create(Array)
#2 /usr/src/public_html/wp-content/plugins/action-scheduler/classes/WP_CLI/Action/Create.php(82): as_enqueue_async_action('woocommerce_cle...', 10)
```

vedanshujain commented on classes/WP_CLI/Action/Generate.php:
```diff
@@ -0,0 +1,124 @@
+		$function_args = array_values( array_filter( $function_args ) );
+
+		try {
+			$actions_added = $this->generate( ...$function_args );
```
let's ensure that we have minimum count of `functions_args` as 4

vedanshujain reviewed (commented):
Posting feedback for some initial testing I was able to get around to today. Seems like we need to sprinkle in some more defensive checks

crstauf committed:
catch deletion failures

crstauf committed:
catch action retrieval errors

crstauf committed:
catch exception when creating action

crstauf committed:
must allow empty function parameters

crstauf committed:
cleanup creation success

crstauf committed:
permit empty function parameters

crstauf committed:
cleanup

crstauf committed:
improve display of action log entries

vedanshujain commented on classes/WP_CLI/Action/Create.php:
```diff
@@ -0,0 +1,137 @@
+			$function    = 'as_schedule_cron_action';
+		}
+
+		$function_args = array_values( $function_args );
```
this is converting a 2d array to 1d, causing issues when we pass this in `call_user_func_array`. So this fails:
` wp action-scheduler action create test_action now`

vedanshujain replied:
Seems like its the same issue -> https://github.com/woocommerce/action-scheduler/pull/1071#discussion_r1666636495

crstauf replied:
Hmm, did not fail for me. What was the output?

vedanshujain replied:
not sure, seems to be working after recent changed, although I don't see anything on first glance that might be causing it.

crstauf committed:
fix version sorting

crstauf committed:
fix 'claimed' parameter

crstauf committed:
explicitly define return format and add to debug info

crstauf committed:
improve debug info

crstauf committed:
cleanup the next command

why was i repeating myself? do this instead:
wp action-scheduler action get $(wp action-scheduler action next hook_name)

crstauf commented:
@barryhughes Stalled here.

barryhughes commented:
Sorry, @crstauf. Please bear with us, as we've had a few other unexpected/unplanned pieces of work demanding our attention, unrelated to this project.

vedanshujain commented on classes/WP_CLI/Action/Run.php:
```diff
@@ -0,0 +1,191 @@
+			 * %1$d: count of actions evaluated.
+			 * %2$s: type of action evaluated.
+			 */
+			$format = _n( '%1$d action %2$s.', '%1$d actions %2$s.', $count, 'action-scheduler' ),
```
syntax error:

```suggestion
			$format = _n( '%1$d action %2$s.', '%1$d actions %2$s.', $count, 'action-scheduler' );
```

vedanshujain reviewed (commented):
Looking good, although seems like something's going on with the cron action:

I tried scheduling a cron action like so:

```
> wp  action-scheduler action create woocommerce_cleanup_draft_orders_test now --cron="1 4 1 3 0"
Success: Async action (67684) scheduled.
```

Above message indicate that action was successfully created, but looking in the DB seems like it's created with a null schedule instead of the cron schedule. See:

<img width="1662" alt="Screenshot 2024-07-29 at 12 18 06 PM" src="https://github.com/user-attachments/assets/64e37a94-7f1e-4927-814a-e3ca48e0b2e2">

I expect it to be similar of following command:

```
>  wp  eval  "as_schedule_cron_action(1, '1 4 1 3 0', 'woocommerce_cleanup_draft_orders_test' );"
```

Which schedules the action like this:

<img width="1648" alt="Screenshot 2024-07-29 at 12 20 51 PM" src="https://github.com/user-attachments/assets/e6c55f07-90b3-4094-a5b3-958ba594b990">


Maybe the values are not being passed as expected?

crstauf committed:
replace comma with semicolon

crstauf committed:
save

crstauf committed:
save

crstauf committed:
add help text

vedanshujain commented:
Thanks for the latest changes, looking pretty good now. I am planning to do bit more testing, but hopefully we will be able to merge this before end of this week.

barryhughes commented on classes/ActionScheduler_AdminView.php:
```diff
@@ -228,6 +228,10 @@ public function add_help_tabs() {
+					'</p>' .
+					'<h3>' . __( 'WP CLI', 'action-scheduler' ) . '</h3>' .
+					'<p>' .
+						__( 'WP CLI commands are available: execute <code>wp help action-scheduler</code> for a list of available commands.' ) .
```
We should escape strings whenever possible, avoid exposing translators to markup, and use text domains. Here's an alternative:

```suggestion
						sprintf(
							/* translators: %1$s contains the WP CLI command 'wp help action-scheduler' (not translatable). */
							esc_html__( 'WP CLI commands are available: execute %1$s for a list of available commands.', 'action-scheduler' ),
							'<code>wp help action-scheduler</code>'
						) .
```

barryhughes commented on classes/ActionScheduler_AdminView.php:
```diff
@@ -228,6 +228,10 @@ public function add_help_tabs() {
 					'<p>' .
 						__( 'Action Scheduler is a scalable, traceable job queue for background processing large sets of actions. Action Scheduler works by triggering an action hook to run at some time in the future. Scheduled actions can also be scheduled to run on a recurring schedule.', 'action-scheduler' ) .
+					'</p>' .
+					'<h3>' . __( 'WP CLI', 'action-scheduler' ) . '</h3>' .
```
```suggestion
					'<h3>' . esc_html__( 'WP CLI', 'action-scheduler' ) . '</h3>' .
```

barryhughes commented on classes/WP_CLI/Action/Cancel.php:
```diff
@@ -0,0 +1,122 @@
+	 * @uses as_unschedule_action()
+	 * @uses as_unschedule_all_actions()
+	 * @uses $this->print_error()
+	 * @uses $this->print_success()
```
We don't typically use these. To reduce noise and keep things readable, I'd recommend we remove them here and elsewhere in this changeset.

```suggestion
```

📝 See [Inline Doc Standards](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/) for WordPress.

barryhughes commented on classes/WP_CLI/Action/Cancel.php:
```diff
@@ -0,0 +1,122 @@
+	 *
+	 * @param string $hook
+	 * @param array $callback_args
+	 * @param string $group
```
Let's make sure we run the linter and get a clean bill of health on all these new files. For instance, we need param comments (you can run `composer run phpcs` to list these things).

barryhughes commented on classes/WP_CLI/Action/Cancel.php:
```diff
@@ -0,0 +1,122 @@
+
+		try {
+			$result = call_user_func( 'as_unschedule_action', $hook, $callback_args, $group );
+		} catch ( \Exception $e ) {
```
Though they don't harm, we're qualifying our use of various types like this one unnecessarily. Here and elsewhere, I think if we can avoid unnecessary symbols, it becomes a little easier to read.

```suggestion
		} catch ( Exception $e ) {
```

crstauf replied:
Going to leave this for now, for consistency (other AS classes use `\Exception`).

barryhughes commented on classes/WP_CLI/Action/Cancel.php:
```diff
@@ -0,0 +1,122 @@
+		\WP_CLI::error(
+			sprintf(
+				/* translators: %1$s: singular or plural %2$s: refers to the exception error message. */
+				__( 'There was an error cancelling the scheduled %1$s: %2$s', 'action-scheduler' ),
```
Preferred approach here (singular/plural versions) is to use `_n()`.

crstauf replied:
There's no numbers available or needed here, which is how `_n()` works. Sure we could force `_n()`, but it's use would be silly in comparison.

barryhughes replied:
Why would it be silly? In English, only the `%1$s` placeholder needs to change. In other languages, preceding words like `scheduled` may need to change, according to their grammar rules. This doesn't allow for that.

barryhughes commented on classes/WP_CLI/Action/Run.php:
```diff
@@ -0,0 +1,191 @@
+	 * @uses \WP_CLI::debug()
+	 * @return void
+	 */
+	public function action__ignored( $action_id ) {
```
Curious, why the double underscore here (and other places)? 

Might a name like `count_ignored_actions` or perhaps `on_action_ignored` be slightly clearer?

crstauf replied:
It's my style, and I find it to be perfectly clear and not overly verbose, especially coupled with the phpdoc.

barryhughes replied:
That's fine, it's not a criticism. However, it's a little atypical for *this* project and our team, though (and we need to strike a balance, and some consistency).

barryhughes commented on classes/WP_CLI/Action/List.php:
```diff
@@ -0,0 +1,135 @@
+				return in_array( $key, static::PARAMETERS, true );
+			},
+			ARRAY_FILTER_USE_KEY
+		);
```
A bit shorter and possibly more efficient to do either:

```suggestion
		$query_args = wp_array_slice_assoc( $this->assoc_args, static::PARAMETERS );
```

Or:

```suggestion
		$query_args = array_intersect_key( $assoc_args, array_flip( static::PARAMETERS ) );
```

barryhughes replied:
Thinking on it some more, I wonder if we should even have this check/filter.

- Should we subsequently add support for new query args, this means we need updates in two different places.
- If a new or custom data store supports additional query args of its own, they won't be supported through this interface.

barryhughes commented on classes/WP_CLI/Action/List.php:
```diff
@@ -0,0 +1,135 @@
+		switch ( $formatter->format ) {
+
+			case 'ids':
+				$actions = as_get_scheduled_actions( $query_args, 'ids' );
```
Optional thing, we could consider moving this to right before we enter the `switch` (instead of repeating it at the start of each `case`).

crstauf replied:
The `default` case is different.

barryhughes commented on classes/WP_CLI/Action/Next.php:
```diff
@@ -0,0 +1,62 @@
+			return;
+		}
+
+		WP_CLI::warning( 'No matching next action.' );
```
Translate.

```suggestion
		WP_CLI::warning( __( 'No matching next action.', 'action-scheduler' ) );
```

crstauf replied:
Core WP CLI commands do not use translation functions:

- https://github.com/wp-cli/entity-command/blob/c270cc9a2367cb8f5845f26a6b5e203397c91392/src/Site_Command.php#L262
- https://github.com/wp-cli/entity-command/blob/c270cc9a2367cb8f5845f26a6b5e203397c91392/src/Post_Command.php#L186

barryhughes commented on classes/WP_CLI/ActionScheduler_WPCLI_System_Command.php:
```diff
@@ -0,0 +1,213 @@
+	 * Construct.
+	 */
+	public function __construct() {
+		$this->store = \ActionScheduler::store();
```
Let's declare this property before using it.

barryhughes commented on classes/WP_CLI/ActionScheduler_WPCLI_System_Command.php:
```diff
@@ -0,0 +1,213 @@
+	}
+
+	/**
+	 * Get latest or all system versions.
```
I wonder if we can make the intent a little clearer?

```suggestion
	 * Display the active Action Scheduler version, or list all registered versions.
```

barryhughes commented on classes/WP_CLI/ActionScheduler_WPCLI_System_Command.php:
```diff
@@ -0,0 +1,213 @@
+		$all      = (bool) get_flag_value( $assoc_args, 'all' );
+		$instance = \ActionScheduler_Versions::instance();
+
+		if ( $all ) {
+			$versions = $instance->get_versions();
+
+			$rows = array();
+
+			foreach ( $versions as $version => $callback ) {
+				$rows[ $version ] = array(
+					'version'  => $version,
+					'callback' => $callback,
+				);
+			}
+
+			uksort( $rows, 'version_compare' );
+
+			$formatter = new \WP_CLI\Formatter( $assoc_args, array( 'version', 'callback' ) );
+			$formatter->display_items( $rows );
+
+			return;
+		}
```
Currently the output is something like this:

```
+---------+-------------------------------------------+
| version | callback                                  |
+---------+-------------------------------------------+
| 3.6.0   | action_scheduler_initialize_3_dot_6_dot_0 |
| 3.8.1   | action_scheduler_initialize_3_dot_8_dot_1 |
+---------+-------------------------------------------+
```

Could it be useful to indicate which is active/loaded?

```
+---------+-------------------------------------------+--------+
| version | callback                                  | active |
+---------+-------------------------------------------+--------+
| 3.6.0   | action_scheduler_initialize_3_dot_6_dot_0 | no     |
| 3.8.1   | action_scheduler_initialize_3_dot_8_dot_1 | yes    |
+---------+-------------------------------------------+--------+
```

We generally hope it will be the latest version, but that is not always so (and my assumption is that this sub-command exists to aid debugging).

crstauf replied:
Oh whoa. What would the situation be where the latest isn't used?

barryhughes replied:
If several versions of Action Scheduler are present and one is 3.2.0 or earlier then, depending on how it is initialized, it may take precedence over later versions [(related)](https://developer.woocommerce.com/2021/10/12/best-practices-for-deconflicting-different-versions-of-action-scheduler/). Not just hypothetical, we came across one such situation quite recently, impacting WooCommerce.

barryhughes replied:
That said, the proposed change could be left until a future iteration.

crstauf replied:
Interesting. The `active` column was added.

barryhughes commented on classes/WP_CLI/Action/Cancel.php:
```diff
@@ -0,0 +1,122 @@
+		}
+
+		try {
+			$result = call_user_func( 'as_unschedule_action', $hook, $callback_args, $group );
```
Since we know the function we're calling, why don't we call it directly (why `call_user_func()`)?

crstauf replied:
I'm not sure. 🤷‍♂️ 😉

barryhughes commented on classes/WP_CLI/ActionScheduler_WPCLI_Action_Command.php:
```diff
@@ -0,0 +1,352 @@
+	 * : Instead of returning the whole action, returns the value of a single field.
+	 *
+	 * [--fields=<fields>]
+	 * : Limit the output to specific fields. Defaults to all fields.
```
```suggestion
	 * : Limit the output to the specified fields (should be comma-separated). Defaults to all fields.
```

barryhughes commented on classes/WP_CLI/Action/Get.php:
```diff
@@ -0,0 +1,72 @@
+			\WP_CLI::error( sprintf( esc_html__( 'Unable to retrieve action %d.', 'action-scheduler' ), $action_id ) );
+		}
+
+		$only_logs   = ! empty( $this->assoc_args['field'] ) && 'log_entries' === $this->assoc_args['field'];
```
Should this also check against `--fields` (ie, if `--fields=log_entries` was passed)?

crstauf replied:
Oh, that's interesting. Why have both `--field` and `--fields` flags? 🤔

barryhughes replied:
Sticking to one may be cleaner, and would avoid doubling up on checks here. Instead of supporting [`--field` and `--fields`](https://github.com/woocommerce/action-scheduler/pull/1071/files#diff-ab2b213587b3f26cf90e53ed137c529583b32100193ae7b256a3c58dbb3d2232R184-R188), we could just rely on `--fields`.

crstauf replied:
Sorry, my question was meant for core. Because core implements both, I'm going to keep both.

barryhughes commented on classes/WP_CLI/Action/Next.php:
```diff
@@ -0,0 +1,62 @@
+			'group'   => $group,
+		);
+
+		if ( is_array( $args ) ) {
+			$params['args'] = $args;
```
Should this be:

```suggestion
		if ( is_array( $this->assoc_args['args'] ) ) {
			$params['args'] = $this->assoc_args['args'];
```

barryhughes commented on classes/abstracts/ActionScheduler_WPCLI_Command.php:
```diff
@@ -0,0 +1,98 @@
+	 * @param array<string, string> $assoc_args Keyed arguments.
+	 */
+	public function __construct( array $args, array $assoc_args ) {
+		if ( ! defined( 'WP_CLI' ) || ! constant( 'WP_CLI' ) ) {
```
```suggestion
		if ( ! defined( 'WP_CLI' ) || ! WP_CLI ) ) {
```

Not sure we need the call to `constant`.

crstauf replied:
Not needed, but more explicit.

barryhughes replied:
Asking out of interest, you mean it makes it more explicit that `WP_CLI` is a constant and not a `$string`? I'd have thought the primary use-case for `constant` is cases where the constant name is unknown (held in a variable) or is formed dynamically.

crstauf replied:
That is likely the intent of `constant()`, but I've found that sometimes PHPStan gets mad at `WP_CLI`, and wrapping it in `constant()` fixes it. Always wondered if it's because there's both a constant and class with the same name.

barryhughes commented on classes/abstracts/ActionScheduler_WPCLI_Command.php:
```diff
@@ -0,0 +1,98 @@
+	public function __construct( array $args, array $assoc_args ) {
+		if ( ! defined( 'WP_CLI' ) || ! constant( 'WP_CLI' ) ) {
+			/* translators: %s php class name */
+			throw new Exception( sprintf( __( 'The %s class can only be run within WP CLI.', 'action-scheduler' ), __CLASS__ ) );
```
`__CLASS__` will be the class we are in (the super-class), not the invoked class. Do we want:

```suggestion
			throw new Exception( sprintf( __( 'The %s class can only be run within WP CLI.', 'action-scheduler' ), get_class( $this ) ) );
```

barryhughes commented on classes/abstracts/ActionScheduler_WPCLI_Command.php:
```diff
@@ -0,0 +1,98 @@
+
+			if ( is_numeric( $recurrence ) ) {
+				/* translators: %s: time interval */
+				return sprintf( __( 'Every %s', 'action-scheduler' ), self::human_interval( $recurrence ) );
```
Can we check on `human_interval`, is this referenced correctly (is this intended to access the `ActionScheduler_ListTable` method)?

crstauf replied:
I suspect I initially copy and pasted it from `ActionScheduler_ListTable` for some reason, but it does not appear that the `get_recurrence()` function is being called.

barryhughes commented on classes/abstracts/ActionScheduler_WPCLI_Command.php:
```diff
@@ -0,0 +1,98 @@
+	 * @link https://github.com/wp-cli/entity-command/blob/6e0e77a297eefa3329b94bec16c15cf7528d343f/src/WP_CLI/CommandWithDBObject.php
+	 * @return void
+	 */
+	protected function process_csv_arguments_to_arrays() {
```
When do we invoke this? I see it is called from `ActionScheduler_WPCLI_Action_List_Command::execute()`, to be clear, but can you give me an example of a command invocation that requires it?

crstauf replied:
I cannot give an example, I only noticed that [several of the core commands](https://github.com/search?q=repo%3Awp-cli%2Fentity-command%20process_csv_arguments_to_arrays&type=code) implement it.

barryhughes replied:
OK, but if in our case we aren't accepting CSV-formatted arguments, why add it?

crstauf replied:
Core's documentation doesn't mention accepting CSV-formatted arguments either, so it's likely they've anticipated some situation that should be accounted for.

If you really want to get rid of it, please do so.

barryhughes reviewed (commented):
It's a big change, thank you for working on it. I added some additional comments, but there's a lot to test here so it may take some more time.

Some high-level thoughts:

- In future, it may be better to break changes like this down into smaller chunks.
- We're adding lots of code here, but no unit tests have been added.
- We're also adding lots of functionality here, but there are no additions to `docs/wp-cli.md` or other doc files. This is something for us to think about.
- We're introducing a new naming convention for class-files. I'm not opposed to it, but am also envisioning a future where we want to introduce PSR-4 conventions, and that could leave us with three different naming conventions.

crstauf committed:
improve translatability

crstauf committed:
remove @uses phpdoc statements

crstauf committed:
remove parameter restriction

crstauf committed:
declare property

crstauf committed:
improve command description

crstauf committed:
add active column to versions table

crstauf committed:
call function directly

crstauf committed:
improve parameter description

crstauf committed:
add check for fields parameter set to log_entries

crstauf committed:
fix variable name

crstauf committed:
change method to retrieve class name

crstauf committed:
remove unused function

crstauf committed:
update link

crstauf committed:
eliminate smelly code

crstauf commented:
> * We're adding lots of code here, but no unit tests have been added.

@barryhughes I don't see any tests for WP CLI commands. Can you please direct me to those so I can reference them? I've never written tests for WP CLI commands before.

crstauf committed:
remove trailing comma

crstauf committed:
add list of commands to documentation

crstauf committed:
eliminate code smells

...and be consistent with \Exception

crstauf committed:
change function names

barryhughes commented on classes/ActionScheduler_AdminView.php:
```diff
@@ -228,6 +228,14 @@ public function add_help_tabs() {
+					'<p>' .
+						sprintf(
+							/* translators: %1$s is WP CLI command (not translatable) */
+							__( 'WP CLI commands are available: execute %1$s for a list of available commands.', 'action-scheduler' ),
```
```suggestion
							esc_html__( 'WP CLI commands are available: execute %1$s for a list of available commands.', 'action-scheduler' ),
```

crstauf committed:
escape translation output

barryhughes commented:
> I don't see any tests for WP CLI commands.

Yep, doesn't look like we have coverage there (so far). We can stick a pin in this aspect while we think about how best to approach it.

crstauf committed:
add todo

crstauf committed:
save

crstauf commented:
@barryhughes @vedanshujain I am unable to reproduce the errors in the unit tests.

![CleanShot 2024-08-09 at 21 06 46@2x](https://github.com/user-attachments/assets/475b9db7-b921-4f3d-8873-c232c13c5157)

barryhughes commented:
Re-running them, in case there was some flakiness.

crstauf commented:
> Re-running them, in case there was some flakiness.

@barryhughes Thanks: looks like all checks passed. 👍

crstauf committed:
Merge branch 'trunk' into wpcli

barryhughes commented:
One thing I'm a bit wary of here is introducing yet another convention in terms of how we name and structure class files. "Traditionally", class files in Action Scheduler seem to be nested in directories representing different sub-systems, and are always prefixed with `ActionScheduler_*`. The class name and the class file are basically 1:1 in these cases. Here's a good example of that:

`classes/data-stores/ActionScheduler_DBLogger.php` contains `ActionScheduler_DBLogger`

We also have started to use namespaces (broadly following PSR-4 conventions), with `classes/` acting as the base directory for the `Action_Scheduler` namespace: 

`classes/WP_CLI/Migration_Command.php` contains `Action_Scheduler\WP_CLI\Migration_Command`

This change introduces yet another convention that doesn't align with either of those:

`classes/WP_CLI/Action/Delete.php` contains `ActionScheduler_WPCLI_Action_Delete_Command`

I think it would be best to align with either of the existing styles, rather than introducing a third approach. Probably, namespacing as in the second example is the best way forward at this time.

crstauf commented:
@barryhughes I'll need to leave that to you, as I'm not interested in fighting with the autoloader for 10 hours.

crstauf committed:
namespace

crstauf committed:
rename files

crstauf committed:
save

crstauf committed:
save

barryhughes commented on classes/WP_CLI/Action/Cancel_Command.php:
```diff
@@ -0,0 +1,120 @@
+		}
+
+		try {
+			$result = call_user_func( 'as_unschedule_all_actions', $hook, $callback_args, $group );
```
I won't mark each one, but as in an earlier thread we should use direct invocation unless there is a reason for `call_user_func` (there doesn't seem to be, because we're passing a string literal and the same set of args each time).

crstauf replied:
This was the last one.

crstauf committed:
save

barryhughes commented on classes/abstracts/ActionScheduler.php:
```diff
@@ -297,9 +300,11 @@ protected static function is_class_migration( $class ) {
+			'QueueRunner'                          => true,
+			'Command'                              => true,
+			'ProgressBar'                          => true,
+			'ActionScheduler_WPCLI_Action_Command' => true,
+			'ActionScheduler_WPCLI_System_Command' => true,
```
Do these need to be updated following the switch to namespacing?

crstauf replied:
@barryhughes I absolutely hate this loading mechanism. May need your help to figure this out.

crstauf replied:
Okay, think I got it.

barryhughes replied:
Thanks. Yep, it's a little tricky to navigate.

barryhughes commented on classes/WP_CLI/ActionScheduler_WPCLI_System_Command.php:
```diff
@@ -0,0 +1,220 @@
+/**
+ * System info WP-CLI commands for Action Scheduler.
+ */
+class ActionScheduler_WPCLI_System_Command {
```
Is it intentional to use this class naming convention here, whereas sibling class `Action_Command` is now namespaced?

crstauf replied:
It was, but I'll change it.

crstauf committed:
save

crstauf committed:
save

crstauf committed:
Merge branch 'master' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf commented:
@barryhughes Stalled again. 😞

barryhughes commented:
Yep, I'm sorry. We've been focusing on other tasks and responsibilities that have taken priority. This is looking great, though, so bear with us a little longer. *Probably,* if any further changes are needed, they'll be minor/small pieces of clean-up, and so we can likely take care of those if required (and can similarly update the branch if it falls out of sync in the meantime).

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

crstauf committed:
Merge branch 'trunk' into wpcli

barryhughes reviewed (approved):
Thanks again, @crstauf. Merging this, and it should therefore become generally available in the next release.

crstauf commented:
Yay! Thanks, @barryhughes!
